### PR TITLE
use Node path from config file to run plugins

### LINKED ISSUE #4162: Native modules used inside plugin fails on Window
### Current behavior:

Using native modules inside the pluginsfile (`plugins/index.js`) will throw `DLL initialization fail` error on Windows. 

### Desired behavior:

Should be able to use native modules inside plugins file

### Steps to reproduce: (app code and test code)

Inside the pluginsfile, I am requiring the `serialport` module. 

```js
const serialport = require('serialport');
```

Below is the error. Also, Cypress was configured to run in Chrome. 

```
Error: A dynamic link library (DLL) initialization routine failed.
\\?\C:\Code\project\node_modules\@serialport\bindings\build\Release\bindings.node
    at process.module.(anonymous function) [as dlopen] (ELECTRON_ASAR.js:172:20)
    at Object.Module._extensions..node (module.js:598:18)
    at Object.module.(anonymous function) [as .node] (ELECTRON_ASAR.js:172:20)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
    at Module.require (module.js:513:17)
    at require (internal/module.js:11:18)
    at bindings (C:\Code\project\node_modules\bindings\bindings.js:112:48)
    at Object.<anonymous> (C:\Code\project\node_modules\@serialport\bindings\lib\win32.js:1:191)
    at Object.<anonymous> (C:\Code\project\node_modules\@serialport\bindings\lib\win32.js:111:3)
    at Module._compile (module.js:569:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
```

### Versions
Cypress Version: `3.1.0`
OS: `Windows 10`
Browser: Chrome `v74.0.3729.131`
Node: `v8.15.1`

### LINKED ISSUE #4432: Allow using external Node when running tasks / plugins
## Proposal

Currently, the plugins file is executed using the Node.js version included with Cypress Test Runner (like 8.2.1), which is often not the version needed to run the application. The most common problem is the inability to load native extensions when requiring NPM modules from the plugins file. The installed app for example uses Node 10 to compile extensions, while Cypress tries to load them using Node 8

So instead of using built-in Node version in `packages/server/lib/plugins/index.coffee`

```
pluginsProcess = cp.fork(path.join(__dirname, "child", "index.js"),
   ["--file", config.pluginsFile], { stdio: "inherit" })
```

we could find external Node and launch it as a child process

## reported issues

- cannot connect to Oracle DB #2914
- cannot load `node-sass` #3875 and #4202

### PR DESCRIPTION
**Pull request by bahmutov**
- closes #4432 
- closes #4162

### Description of change

Added a config option `nodeVersion` that will be used to determine which version of Node Cypress is run within. Users can use the `bundled` version that comes with Cypress (which is the default) or override it with the `system` Node version installed on their machine.

### How to use the feature

In configuration file (cypress.json by default), to use the bundled version of Node with Cypress. `default` or `bundled` = use the Node that comes with Cypress Test Runner

```json
{
  "nodeVersion": "bundled"
}
```

In configuration file (cypress.json by default), to use the system version of Node with Cypress. `system` = tries to find system Node using cross-platform https://github.com/npm/node-which from NPM. If fails to find, tries again but first fixing path using https://github.com/sindresorhus/fix-path

```json
{
  "nodeVersion": "system"
}
```

### How the design has changed

- [x] add frontend view in "Settings" to indicate path to node + node version in use
- [x] update `cypress run` mode to indicate the system version of node (likely in the table header at the top of a run shown below)

```diff
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        1.2.3                                                                          │
  │ Browser:        FooBrowser 88                                                                  │
  │ Node Version:   v0.0.0 (bundled with Cypress)                                                  │
  │ Specs:          1 found (base_url_spec.coffee)                                                 │
  │ Searched:       cypress/integration/base_url_spec.coffee                                       │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```

To demo / test this, I made a branch `external-node-for-plugins-4432` in https://github.com/cypress-io/cypress-test-tiny/tree/external-node-for-plugins-4432 that runs `cy.task` that prints Node version. To execute from the repo root run (assuming the test project is in `~/git/cypress-test-tiny/`). Example output from the terminal running Node 12

```text
$ npm run dev -- --run-project ~/git/cypress-test-tiny/

> cypress@3.4.0 dev /Users/gleb/git/cypress
> node ./scripts/start.js "--run-project" "/Users/gleb/git/cypress-test-tiny/"

plugins file Node version v12.4.0
plugins file on, Node version v12.4.0

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    3.4.0                                                                              │
  │ Browser:    Electron 61 (headless)                                                             │
  │ Specs:      1 found (spec.js)                                                                  │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running: spec.js...                                                                      (1 of 1) 
file:preprocessor file /Users/gleb/git/cypress-test-tiny/cypress/integration/spec.js Node version v12.4.0
file:preprocessor file /Users/gleb/git/cypress-test-tiny/cypress/support/index.js Node version v12.4.0
```

#### New Node Version panel in Desktop GUI

**Using node version bundled with Cypress**

Screen Shot 2019-10-01 at 1 44 47 PM

**Using node version from your system**

Screen Shot 2019-10-01 at 1 46 11 PM

### Pre-merge Tasks

- [x] docs PR: cypress-io/cypress-documentation#1876
- [x] add tests

### CODE REVIEW

**Comment by bahmutov** [comment]
hmm, maybe it is better to find Node using https://github.com/npm/node-which

**Comment by brian-mann** [comment]
@bahmutov 	you should use this module: https://github.com/sindresorhus/run-node

If Cypress is run in global "double click" mode then `$PATH` will not be available. Sindre's package will effectively *always* find the correct node version correctly - and it's used by things like husky git hooks which work in git gui apps.

**Comment by bahmutov** [comment]
I looked at that module and I am not sure it will work on Windows at all

**Comment by bahmutov** [comment]
plus `run-node` is just a bash shell script to run Node - we need to find the path to the node, so at most we can reuse some logic from that shell script. I think I can lift this logic from execa though - load user's `$PATH` before using `node-which` is is cross platform

**Comment by bahmutov** [comment]
Other relative utilities:
- https://github.com/sindresorhus/npm-run-path
- https://github.com/sindresorhus/fix-path

**Comment by bahmutov** [comment]
Example (after building and opening Electron binary on Mac)

```
$ nvm use 8
Now using node v8.9.4 (npm v6.9.0)
~/git/cypress on external-node-for-plugins-4432*
$ DEBUG=cypress:server:plugins ./build/darwin/Cypress.app/Contents/MacOS/Cypress 
  cypress:server:plugins plugins.init /Users/gleb/git/cypress-test-tiny/cypress/plugins/index.js +0ms
  cypress:server:plugins finding Node with $PATH /Library/Frameworks/Python.framework/Versions/2.7/bin:/opt/local/bin:/opt/local/sbin:/Users/gleb/.nvm/versions/node/v8.9.4/bin:/opt/local/bin:/opt/local/sbin:/opt/local/bin:/opt/local/sbin:/usr/local/bin:/Users/gleb/dev/arcanist/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Applications/Postgres.app/Contents/Versions/latest/bin +0ms
  cypress:server:plugins found Node /Users/gleb/.nvm/versions/node/v8.9.4/bin/node +0ms
  cypress:server:plugins using Node /Users/gleb/.nvm/versions/node/v8.9.4/bin/node +1ms
plugins file Node version v8.9.4
plugins file on, Node version v8.9.4
```

Then use Node 6
```
$ nvm use 6
Now using node v6.12.2 (npm v3.10.10)
~/git/cypress on external-node-for-plugins-4432*
$ DEBUG=cypress:server:plugins ./build/darwin/Cypress.app/Contents/MacOS/Cypress 
  cypress:server:plugins plugins.init /Users/gleb/git/cypress-test-tiny/cypress/plugins/index.js +0ms
  cypress:server:plugins finding Node with $PATH /Library/Frameworks/Python.framework/Versions/2.7/bin:/opt/local/bin:/opt/local/sbin:/Users/gleb/.nvm/versions/node/v6.12.2/bin:/opt/local/bin:/opt/local/sbin:/opt/local/bin:/opt/local/sbin:/usr/local/bin:/Users/gleb/dev/arcanist/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Applications/Postgres.app/Contents/Versions/latest/bin +0ms
  cypress:server:plugins found Node /Users/gleb/.nvm/versions/node/v6.12.2/bin/node +1ms
  cypress:server:plugins using Node /Users/gleb/.nvm/versions/node/v6.12.2/bin/node +2ms
plugins file Node version v6.12.2
plugins file on, Node version v6.12.2
```

**Comment by bahmutov** [comment]
before I go to write some tests, is this ok for this feature? seems to be enough to control the Node version.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -374,3 +374,20 @@ We use [RenovateBot](https://renovatebot.com/) to automatically upgrade our depe
 We will try to review and merge pull requests quickly. After merging we
 will try releasing a new version. If you want to know our build process or
 build your own Cypress binary, read [DEPLOY.md](DEPLOY.md)
+
+## Known problems
+
+### ENFILE
```
Comment (brian-mann) [inline CONTRIBUTING.md]: I believe you mean `EMFILE`, not `ENFILE`.

This should never happen as long as we *never* use synchronous FS methods - which I'm pretty sure we don't. This is what `graceful-fs` fixes.

If you are seeing this it means we messed something up and need to understand where it's happening and fix it.

Comment (bahmutov) [reply CONTRIBUTING.md]: this was during the build

Comment (bahmutov) [reply CONTRIBUTING.md]: turns out to be sync file system bug - #4448

**Comment by brian-mann** [comment]
Can we rename `"node": "find"` to something like `"node": "*"` or like `<system>` - something that doesn't look like a path that helps communicate that we're going to use the current one enabled in the shell?

Another idea: we could automatically detect if you're using a `.node-version` in your project (or other common files) and attempt to source that one in too.

Why? Because if you're running on multiple projects you may switch the node version between shells, which would then cause Cypress to pick the "last one". Ideally if there's something **in** your project that specifies it, that would be the best way for us to know which one to switch to.

Another idea: provide us the actual version like `"node": "10.0.0"` which will fail if it's not installed via `nvm`, etc. Then it's super clear.

**Comment by brian-mann** [comment]
Beyond the use of plugins - the big win of this is that when we process your spec files, that we use **your** version of `node`.

I believe so far you've demonstrated switching node only for plugins but not for spec processing.

**Comment by brian-mann** [comment]
Thinking about this a bit more... I actually *dont* think we should support absolute paths to the node version. Nothing in `cypress.json` should be user specific - every setting should be OS independent and compatible.

I guess we could still allow for it but maybe display a warning. A better approach is anchoring ourselves to something in the project files, or the user providing a specific node version so we know when it's correct and when it's not correct.

**Comment by brian-mann** [comment]
Bleh... I guess this would likely making running in CI more difficult, which is why the `*` is likely necessary, but not recommended.

Maybe something like how heroku does it - enabling you to pass in a node version like `10` or `10.1` or `10.1.2`.

**Comment by bahmutov** [comment]
so I would love to specify Node version (and read it from `.node-version` if found) like "*" to find any version, and "10.0.0" to use specific, but then we need to figure out how to find it reliably cross-platform. Which is harder than finding current Node and falling back to `fix-path` + find again if it fails.

Then the second iteration could be "find specific version of Node"

I will look into bundling now

**Comment by brian-mann** [comment]
I think the challenge here is that whatever they specify may be problematic when running in CI, etc.

Could we just download that version of node if we can't find it locally and like... cache it in the Cypress cache folder for future use?

I'm also imagining a UI in the config area that gives you a series of radios to help you pick the node version, and whether or not it was found on your system.

Honestly downloading the specific version of node would be a pretty failsafe way of always having it in all environments and users.

**Comment by bahmutov** [comment]
Automatic downloading to me looks almost magical, even if we cache it into `~/.cache` automatically. I feel like most people would just love to use whatever system version of Node they have right now - which could even be different on local machine vs CI - as long as it is not the Node version that comes with Cypress. I for example would love to use Node 10 to run plugins (node-sass) locally, because Parcel bundler requires it, and on CI would just use Docker image `cypress/base:10`

**Comment by bahmutov** [comment]
I have checked the bundling of files - it uses the found Node version, just like processing of plugins file. In the log below, the first Node version if the Cypress Test Runner own code (Node 8) and then the bundling prints its Node version (12)

```
  cypress:server:preprocessor getFile /Users/gleb/git/cypress-test-tiny/cypress/integration/spec.js +0ms
  cypress:server:plugins plugin event registered? { event: 'file:preprocessor', isRegistered: true } +1s
  cypress:server:plugins execute plugin event 'file:preprocessor' Node 'v8.9.3' with args: EventEmitter { domain: null, _events: { rerun: [Function] }, _eventsCount: 1, _maxListeners: undefined, filePath: '/Users/gleb/git/cypress-test-tiny/cypress/integration/spec.js', shouldWatch: true, outputPath: '/Users/gleb/Library/Application Support/Cypress/cy/development/projects/cypress-test-tiny-a98bbf8a7f1688a4442c8bc398984ac5/bundles/cypress/integration/spec.js' } undefined undefined +0ms
  cypress:server:plugins call event file:preprocessor for invocation id inv2 +2ms
file:preprocessor file /Users/gleb/git/cypress-test-tiny/cypress/integration/spec.js Node version v12.4.0
```

**Comment by bahmutov** [comment]
Ok, after thinking some more, I believe the great majority of issues related to the built-in Node version can be solved by loading system Node. For this I think we should allow just two options in `cypress.json`

- `"node": "built-in"` runs the bundled Node, which is the only possible behavior today
- `"node": "system"` should find the current OS Node and use it to execute the plugins file

I think any more features (like UI drop-down, auto-installing Node, trying to read `.node-version` and find the specific Node version) are nice, but are a lot more work and not strictly necessary to solve native extensions and JavaScript features in plugins file problems.

**Comment by brian-mann** [comment]
You are probably right... but maybe there is some middle ground.

At the very least we need to:

- add the node version to the output via `cypress run` and likely need to capture it and display it in the dashboard
- display the node version in the Settings tab via `cypress open` so you can at least *see* which version it's using and where it found that version of node.

The property for this should likely be: `nodeVersion` and if we did just stick to two options it would probably be...

- `default` or `bundled`
- `system`

**Comment by bahmutov** [comment]
yeah, agree that we need to show it in the `cypress run` output and in the GUI but I would do it in the separate issue. For now I will clean this PR to remove relative user path stuff and update tests

**Comment by bahmutov** [comment]
bump, had to jump through hoops in https://github.com/bahmutov/realworld/blob/master/package.json#L20 to have Node8 native bindings for sqlite3 to work while serving app with Parcel bundler that requires Node10

**Comment by brian-mann** [comment]
I've updated this PR to include updating the frontend in "Settings" and to update `cypress run` mode to indicate which system node version + path is in use.

**Comment by flotwig** [comment]
### Bundled
image

### Custom
image

### Custom w/ failed version check
image

**Comment by brian-mann** [comment]
> ### Bundled
> image
> 
> ### Custom
> image
> 
> ### Custom w/ failed version check or could not find `node`
> image
> 
> image

When using the bundled node version - drop the `Node Path:` label and just use this...

```
Node Version: v1.2.3 (bundled with Cypress)
```

### TODO / other notes / thoughts
We should design the default node version view differently than it is now. Users didn't **do** anything, and we should use the space as an opportunity to educate that you **can** change the node version, but additionally explain **why** you'd want to do it.

At the very least we should provide a link to a doc that goes into more detail.

**Comment by flotwig** [comment]
Updated desktop-gui and added a Learn More link (to https://on.cypress.io/configuration#Node-Version for now): 
image

Updated run mode so that it only prints out the Node version when it's non-bundled.

File: packages/server/lib/util/find_system_node.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    throw err
+  })
+}
+
+function findNodeVersionFromPath (path) {
+  if (!path) {
+    return Promise.resolve(null)
+  }
+
+  return execa.stdout(path, ['-v'])
```
Comment (bahmutov) [inline packages/server/lib/util/find_system_node.js]: where is `execa.stdout` coming from?

Comment (flotwig) [reply packages/server/lib/util/find_system_node.js]: We're still on `execa@1.0.0`: https://github.com/sindresorhus/execa/tree/v1.0.0#execastdoutfile-arguments-options

Comment (bahmutov) [reply packages/server/lib/util/find_system_node.js]: ok, thought so too

**Comment by jennifer-shehane** [comment]
So, honestly this is more of a documentation issue than design - and I need some things answered about what this does. 

1. Does the Node version affect the code run in `cy.exec()`
2. Does the Node version affect the code run in `support/*` files?

**Comment by jennifer-shehane** [comment]
Screen Shot 2019-07-18 at 3 12 19 PM

**Comment by flotwig** [comment]
@jennifer-shehane 

1. `cy.exec` will still be executed in Cypress's bundled node version
2. I think the `support/` files are all loaded in-browser, so this won't affect that

**Comment by jennifer-shehane** [comment]
Second iteration on wording of Node version panel.

Screen Shot 2019-07-19 at 10 23 19 PM

**Comment by flotwig** [comment]
I feel like the first sentence is very confusing. "used to build files found in cypress/integration and cypress/support" is very obscure, especially since by default, they don't need to put anything in their pluginsFile to make this happen. 

Maybe it would be better to keep the last two sentences and just to link to this page in the docs which explains what the pluginsFile does? https://docs.cypress.io/guides/tooling/plugins-guide.html

**Comment by jennifer-shehane** [comment]
More design, I'm very open to better wording suggestions. 

Screen Shot 2019-07-19 at 10 46 03 PM

**Comment by jennifer-shehane** [comment]
K, newest design + fixed tests @brian-mann 	

**When using bundled version**

Screen Shot 2019-07-25 at 4 47 34 PM

**When using system version**

Screen Shot 2019-07-25 at 4 48 12 PM

**Comment by brian-mann** [comment]
Some new designs...

### Bundled

Screen Shot 2019-07-30 at 12 24 08 AM

### System

Screen Shot 2019-07-30 at 12 20 18 AM

**Comment by brian-mann** [comment]
When the version is set to the default `bundled`, I would like to include a trailing suffix in the well that explains the **why**. The bullets are good, but they only explain the **what**.

**Comment by jennifer-shehane** [comment]
Updated. I didn't add a **why** because I'm not sure exactly what copy you want written there.

File: packages/desktop-gui/src/settings/node-version.jsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    <td>
+                      <code>{resolvedNodePath}</code>
+                    </td>
+                  </tr>
+                </tbody>
+              </table>
+              <p className="text-muted node-version-text">
+                <strong>Note:</strong> This version of Node.js was detected on your{' '}
+                <a href="#" onClick={openHelp}>
+                  <code>system</code>
```
Comment (flotwig) [inline packages/desktop-gui/src/settings/node-version.jsx]: `system` being in `<code>` doesn't make sense to me. I guess it's supposed to reference that `nodeVersion` can be `system`, but I didn't immediately make that connection.

Comment (jennifer-shehane) [reply packages/desktop-gui/src/settings/node-version.jsx]: yeah 😬

**Comment by mpahuja** [comment]
Any ETA on this PR. Would love to get this support in for the use case that i am trying to solve for where in the application code and some of the dependencies are built for node@10 and above only. And I want to use the same webpack config for cypress as for the actual application. 
So would love to get this functionality merged in sooner than later.

**Comment by mpahuja** [comment]
in the meantime is there a workaround available to be able to load (using babel-loader) a package built for a different version of node? I tried looking and most of the solutions have been to avoid using such a packet

**Comment by bahmutov** [comment]
@mpahuja, unfortunately, the change is inside the Cypress compiled binary code, so we cannot even release NPM package fix. For me the work here is done, it is up to @brian-mann to review and merge it (soon)

**Comment by mpahuja** [comment]
I am sure and also hopeful that he will get to it as soon as he can.

**Comment by mpahuja** [comment]
@bahmutov what do you think about using this https://github.com/electron/electron-rebuild as a workaround in the meantime?

**Comment by drumbeg** [comment]
> @bahmutov what do you think about using this https://github.com/electron/electron-rebuild as a workaround in the meantime?

Try it! It didn't work consistently for me whilst trying to get the Oracle Node driver working on Windows. Seems to depend on exact set-up.

**Comment by bahmutov** [comment]
yeah @mpahuja I think just updating our code by merging this pull request is a lot better way than rebuilding native modules.

**Comment by mpahuja** [comment]
completely understand Gleb, just curious on the timeline for this to be merged so that I can make an informed decision about whether i need to use the workaround or not.

**Comment by jennifer-shehane** [comment]
This test is failing because when run locally - the node version is 8.0.0 then in CI it is 12.0.0, but we are blindly replacing `8.0.0` or `12.0.0` with `0.0.0` so that there is an extra char being lost in CI. I need to update the snapshot replacement to account for and replace the exact number of digits from Node version.

File: packages/server/lib/modes/run.js [hunk 4]
```diff
@@ -95,6 +95,15 @@ const formatFooterSummary = function (results) {
   ]
 }
 
+const formatNodeVersion = ({ resolvedNodeVersion, resolvedNodePath }) => {
+  debug('formatting Node version. Resolved version "%s", path if any "%s"',
```
Comment (brian-mann) [inline packages/server/lib/modes/run.js]: Use `%o` with key/value instead of `%s` so there's more context as to where the values come from.

**Pull request merged**